### DISCUSSION TITLE
Practice Exercise - React and Testing - Jest - Supermarket Management System

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/mern#186] [Tasks]
Let's create testing scenarios for the Supermarket Management System app in discussion #185 that we discussed before.

## Part 1: Setting Up Jest and React Testing Library

Assuming that Jest and React Testing Library are installed in your project. If not, install them:

```bash
npm install --save-dev jest @testing-library/react @testing-library/user-event
```

## Part 2: Writing Tests

We'll write tests for each component.

### 1. Testing the `App` component

The `App` component is the parent component in our application. It contains the Router and routes to different pages.

```jsx
// App.test.js
import { render } from '@testing-library/react';
import App from './App';

test('renders correct routes', () => {
  const { getByText } = render(<App />);
  expect(getByText('Login')).toBeInTheDocument();
  expect(getByText('Sign Up')).toBeInTheDocument();
  expect(getByText('Products')).toBeInTheDocument();
});
```

This test verifies that the `App` component renders correct routes.

### 2. Testing the `Login` component

The `Login` component provides a form for user authentication.

```jsx
// Login.test.js
import { render, fireEvent } from '@testing-library/react';
import Login from './Login';

test('allows the user to login successfully', async () => {
  const fakeUserResponse = { token: 'fake_user_token' };
  jest.spyOn(window, 'fetch').mockImplementationOnce(() => {
    return Promise.resolve({
      json: () => Promise.resolve(fakeUserResponse),
    });
  });

  const { getByLabelText, getByText } = render(<Login />);

  fireEvent.change(getByLabelText(/username/i), {
    target: { value: 'chris' },
  });
  fireEvent.change(getByLabelText(/password/i), {
    target: { value: 'password' },
  });

  fireEvent.click(getByText(/submit/i));

  const alert = await window.alert;
  expect(alert).toEqual('Logged in!');
});
```

This test verifies that the `Login` component allows the user to log in successfully.

### 3. Testing the `SignUp` component

The `SignUp` component allows new users to create an account.

```jsx
// SignUp.test.js
import { render, fireEvent } from '@testing-library/react';
import SignUp from './SignUp';

test('allows the user to sign up successfully', async () => {
  const fakeUserResponse = { token: 'fake_user_token' };
  jest.spyOn(window, 'fetch').mockImplementationOnce(() => {
    return Promise.resolve({
      json: () => Promise.resolve(fakeUserResponse),
    });
  });

  const { getByLabelText, getByText } = render(<SignUp />);

  fireEvent.change(getByLabelText(/username/i), {
    target: { value: 'chris' },
  });
  fireEvent.change(getByLabelText(/password/i), {
    target: { value: 'password' },
  });

  fireEvent.click(getByText(/submit/i));

  const alert = await window.alert;
  expect(alert).toEqual('Account created!');
});
```

This test verifies that the `SignUp` component allows the user to sign up successfully.

### 4. Testing the `Products` component

The `Products` component displays a list of products.

```jsx
// Products.test.js
import { render } from '@testing-library/react';
import Products from './Products';

test('renders products', () => {
  const products = [
    { id: 1, name: 'Product 1', price: 100 },
    { id: 2, name: 'Product 2', price: 200 },
  ];
  const { getByText } = render(<Products products={products} />);

  expect(getByText('Product 1')).toBeInTheDocument();
  expect(getByText('Product 2')).toBeInTheDocument();
});
```

This test verifies that the `Products` component renders products correctly.

## Part 3: Running the Tests

You can now run your tests using the following command:

```bash
npm test
```

You should see a message indicating that your tests passed!

These are just basic tests, but they should give you a good starting point for testing your Supermarket Management System. Remember, testing is a crucial part of development that ensures your code works as expected. As your application grows, you should aim to cover more scenarios with your tests.

**Comment by sam2elect** [comment]
# TEST
![image](https://github.com/akash-coded/mern/assets/85131631/6071ffac-7335-4b5e-977d-435e77a8a120)
![image](https://github.com/akash-coded/mern/assets/85131631/783705d8-2aa9-4272-8a01-a2fafa65570d)
![image](https://github.com/akash-coded/mern/assets/85131631/84bca550-c9be-4595-a48f-3ebc32b0d48a)

**Comment by KailashGareja** [comment]
![image](https://github.com/akash-coded/mern/assets/136106221/7e0d4a67-bb0d-40db-9a6e-91ad68114223)

**Comment by mightyrsm** [comment]
<img width="1017" alt="Screenshot 2023-08-06 at 10 55 16 PM" src="https://github.com/akash-coded/mern/assets/136921146/0911b6fc-6e7b-46a0-8092-54c29f5e7ca4">

Rahul this side.
I have managed to get 1 test to pass as attached.

Thanks
